### PR TITLE
cmd/snap: implemented add, remove, purge, refresh, rollback, activate, deactivate

### PR DESCRIPTION
**Pull request by chipaca**

### CODE REVIEW

File: cmd/snap/cmd_snap_op_test.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+			fmt.Fprintln(w, `{"type": "sync", "result": {"status": "succeeded"}}`)
+		default:
+			c.Fatalf("expected to get 3 requests, now on %d", n)
+		}
+
+		n++
+	})
+	rest, err := snap.Parser().ParseArgs([]string{"add", "foo.bar"})
+	c.Assert(err, check.IsNil)
+	c.Assert(rest, check.DeepEquals, []string{})
```
Comment (zyga) [inline cmd/snap/cmd_snap_op_test.go]: How about checking that s.Stdout() and s.Stderr() are both empty? It's harmless now and can catch interesting stuff later.

File: cmd/snap/cmd_snap_op.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+`)
+
+var longRollbackHelp = i18n.G(`
+The rollback command reverts an installed snap to its previous revision.
+`)
+
+var longActivateHelp = i18n.G(`
+The activate command activates an installed but inactive snap.
+
+Non-active snaps' apps are not available for use.
```
Comment (zyga) [inline cmd/snap/cmd_snap_op.go]: Non-active sounds a bit weird to me (but feel free to leave it as is). How about:

```Snaps that are not active don't have their apps available for use.```

File: cmd/snap/cmd_snap_op.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+var longDeactivateHelp = i18n.G(`
+The deactivate command deactivates an installed, active snap.
+
+Deactivated snaps' apps are not available for use.
+`)
+
+type cmdOp struct {
+	Positional struct {
+		Snap string `positional-arg-name:"snap"`
+	} `positional-args:"yes"`
```
Comment (zyga) [inline cmd/snap/cmd_snap_op.go]: Don't we want required:"yes" here?

File: cmd/snap/cmd_snap_op_test.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+	"fmt"
+	"net/http"
+
+	"gopkg.in/check.v1"
+
+	snap "github.com/ubuntu-core/snappy/cmd/snap"
+)
+
+func (s *SnapSuite) TestAdd(c *check.C) {
+	n := 0
```
Comment (zyga) [inline cmd/snap/cmd_snap_op_test.go]: Ahhh, I love this. Nice :)

File: cmd/snap/cmd_snap_op.go [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+func init() {
+	for _, s := range []struct {
+		name  string
+		short string
+		long  string
+		op    func(*client.Client, string) (string, error)
+	}{
+		{"add", shortAddHelp, longAddHelp, (*client.Client).AddSnap},
+		{"remove", shortRemoveHelp, longRemoveHelp, (*client.Client).RemoveSnap},
```
Comment (zyga) [inline cmd/snap/cmd_snap_op.go]: Do you plan to test all of those commands as well?

**Comment by zyga** [comment]
Nothing blocking, we could always use more tests ;-). I'm glad the simple CLI testing primitives I've built are working.

+1

File: cmd/snap/cmd_snap_op.go (hunk 3 above)
Comment (chipaca) [inline cmd/snap/cmd_snap_op.go]: good point; yes

File: cmd/snap/cmd_snap_op.go (hunk 2 above)
Comment (chipaca) [inline cmd/snap/cmd_snap_op.go]: :+1:

File: cmd/snap/cmd_snap_op.go (hunk 5 above)
Comment (chipaca) [inline cmd/snap/cmd_snap_op.go]: not at this level, i don't think? not until they're more different at least

Comment (zyga) [inline cmd/snap/cmd_snap_op.go]: Thanks for confirming. I think that's fine.

File: cmd/snap/cmd_snap_op.go [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+var longDeactivateHelp = i18n.G(`
+The deactivate command deactivates an installed, active snap.
+
+Snaps that are not active don't have their apps available for use.
+`)
+
+type cmdOp struct {
+	Positional struct {
+		Snap string `positional-arg-name:"snap"`
```
Comment (pedronis) [inline cmd/snap/cmd_snap_op.go]: seems it should be <snap> ?

Comment (zyga) [inline cmd/snap/cmd_snap_op.go]: Or even <snap name> for consistency across commands.

Comment (chipaca) [inline cmd/snap/cmd_snap_op.go]: we don't use `<foo>` anywhere else, so why does it seem to you we should use it here?

Comment (zyga) [inline cmd/snap/cmd_snap_op.go]: We use <..> in grant/revoke/skills

File: cmd/snap/cmd_snap_op.go [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+	for _, s := range []struct {
+		name  string
+		short string
+		long  string
+		op    func(*client.Client, string) (string, error)
+	}{
+		{"add", shortAddHelp, longAddHelp, (*client.Client).AddSnap},
+		{"remove", shortRemoveHelp, longRemoveHelp, (*client.Client).RemoveSnap},
+		{"purge", shortPurgeHelp, longPurgeHelp, (*client.Client).PurgeSnap},
+		{"refresh", shortRefreshHelp, longRefreshHelp, (*client.Client).RefreshSnap},
```
Comment (pedronis) [inline cmd/snap/cmd_snap_op.go]: wasn't the plan to also have  a refresh that refreshes everything? will that come later?

File: cmd/snap/cmd_snap_op.go [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+		op, err := client.Operation(uuid)
+		if err != nil {
+			return err
+		}
+
+		if !op.Running() {
+			return op.Err()
+		}
+
+		time.Sleep(100 * time.Millisecond)
```
Comment (pedronis) [inline cmd/snap/cmd_snap_op.go]: mmh, should we print at least something after a long time?

**Comment by pedronis** [comment]
looks good, mostly added a couple of wonderings/questions

File: cmd/snap/cmd_snap_op.go (hunk 7 above)
Comment (chipaca) [inline cmd/snap/cmd_snap_op.go]: yes; that's when these start to diverge. I know at least `refresh` and `add` will need to be separate.

File: cmd/snap/cmd_snap_op.go (hunk 8 above)
Comment (chipaca) [inline cmd/snap/cmd_snap_op.go]: nah. We'll get progress bars.

**Pull request merged**